src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3125,8 +3125,28 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                                       ASR::down_cast<ASR::List_t>(type)->m_type, nullptr);
                 return false;
             } else if (ASR::is_a<ASR::Tuple_t>(*type)) {
+                int i = 0;
                 index = ASRUtils::EXPR(tmp);
-                int i = ASR::down_cast<ASR::IntegerConstant_t>(ASRUtils::EXPR(tmp))->m_n;
+                ASR::expr_t* val = ASRUtils::expr_value(index);
+                if (!val) {
+                    throw SemanticError("Runtime Indexing with " + ASRUtils::type_to_str_python(type) +
+                    " is not possible.", loc);
+                }
+
+                if (ASR::is_a<ASR::IntegerConstant_t>(*val)) {
+                    i = ASR::down_cast<ASR::IntegerConstant_t>(val)->m_n;
+                    int tuple_size =  ASR::down_cast<ASR::Tuple_t>(type)->n_type;
+                    if (i < 0) {
+                        i = tuple_size + i;
+                        ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                                                        4, nullptr, 0));
+                        index = ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                            al, loc, i, int_type));
+                    }
+                    if (i >= tuple_size || i < 0) {
+                        throw SemanticError("Tuple index out of bounds", loc);
+                    }
+                }
```

Comment:
What happens if it is not an IntegerConstant? Shouldn't we give a compiler error then?

Reply:
Yeah , we've discussed that in this thread (https://github.com/lcompilers/lpython/pull/1534#discussion_r1113364782) and raised an error on line `3132` for it ! So I think this has been addressed !?

Reply:
It could still be a compile time constant, such as a Real: `t[-1.0]` or Character `t["x"]`, which I think would go through, but we should give an error message.

Reply:
I've added an else block for this , raising an appropriate error message !

Reply:
Yes, this is good now. We don't want to have any unchecked branches in the code, always give an error message or handle it, but we should not be ignoring things.